### PR TITLE
cilium, gops: remap to fixed port to avoid collision with nodeport range

### PR DESCRIPTION
**Pull request by borkmann**
See commit msg.

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Review by rolinh — CHANGES_REQUESTED** [review]
Nice catch! Could you please also update Hubble Relay as I think it's also affected (nodeport protection is global)? It also enables gops by default with the default listen address (see `hubble-relay/cmd/serve/serve.go`).

**Comment by borkmann** [comment]
> Nice catch! Could you please also update Hubble Relay as I think it's also affected (nodeport protection is global)? It also enables gops by default with the default listen address (see `hubble-relay/cmd/serve/serve.go`).

Ah, good point, that one fell through the cracks :/ will add.

**Review by aanm — CHANGES_REQUESTED** [review]
the overall changes LGTM, however a couple points that we need to take care:
1) It would likely be better to `Fatal` if we can't initialize gops. If gops isn't running at all we won't be able to debug Cilium if we need it;
2) Currently, the default is `127.0.0.1:0` and we are changing it to `localhost:XXXX`, maybe we should keep the `127.0.0.1`.
3) I don't see a "reuse port" in gops, so it is likely that we will hit something similar as https://github.com/cilium/cilium/issues/11573.

**Comment by borkmann** [comment]
> the overall changes LGTM, however a couple points that we need to take care:
> 
> 1. It would likely be better to `Fatal` if we can't initialize gops. If gops isn't running at all we won't be able to debug Cilium if we need it;

Ok, fair enough, I'll change to error out with Fatal instead.

> 2. Currently, the default is `127.0.0.1:0` and we are changing it to `localhost:XXXX`, maybe we should keep the `127.0.0.1`.

No strong opinion here, I can change it back. Locally, I had it on `[::1]:XXX` first to map it to v6 only to avoid occupying a v4 port, but then I thought `localhost` instead would be cleaner to not make any addr family assumptions.

> 3. I don't see a "reuse port" in gops, so it is likely that we will hit something similar as #11573.

I haven't run into it on local testing at least and looking into gops code it seems to gracefully clean up the socket on termination, but maybe @tklauser has more insight as gops contributor. :)

File: hubble-relay/cmd/serve/serve.go [hunk 1]
```diff
@@ -184,7 +189,8 @@ func runServe(vp *viper.Viper) error {
 	}
 	gopsEnabled := vp.GetBool(keyGops)
 	if gopsEnabled {
-		if err := agent.Listen(agent.Options{}); err != nil {
+		addr := fmt.Sprintf("localhost:%d", viper.GetInt(keyGopsPort))
```
Comment (rolinh) [inline hubble-relay/cmd/serve/serve.go]: This won't work as there is no global viper instance and thus the port will always be `0`. Please, use `vp` instead.
```suggestion
		addr := fmt.Sprintf("localhost:%d", vp.GetInt(keyGopsPort))
```

**Review by rolinh — CHANGES_REQUESTED** [review]
The Hubble Relay change is not correct, please see comment.

File: hubble-relay/cmd/serve/serve.go (hunk 1 above)
Comment (borkmann) [reply hubble-relay/cmd/serve/serve.go]: Indeed, thanks a lot for the catch!

**Review by rolinh — APPROVED** [review]

**Comment by borkmann** [comment]
test-me-please

**Comment by tklauser** [comment]
> > the overall changes LGTM, however a couple points that we need to take care:
> > 
> > 1. It would likely be better to `Fatal` if we can't initialize gops. If gops isn't running at all we won't be able to debug Cilium if we need it;
> 
> Ok, fair enough, I'll change to error out with Fatal instead.
> 
> > 1. Currently, the default is `127.0.0.1:0` and we are changing it to `localhost:XXXX`, maybe we should keep the `127.0.0.1`.
> 
> No strong opinion here, I can change it back. Locally, I had it on `[::1]:XXX` first to map it to v6 only to avoid occupying a v4 port, but then I thought `localhost` instead would be cleaner to not make any addr family assumptions.

FWIW, we had #13165 where the agent health endpoint was listening on `localhost:` and this lead to problems in IPV6-only environments. Not quite sure whether that is applicable here, but I think sticking to `127.0.0.1` is the safer option.

> > 1. I don't see a "reuse port" in gops, so it is likely that we will hit something similar as #11573.
> 
> I haven't run into it on local testing at least and looking into gops code it seems to gracefully clean up the socket on termination, but maybe @tklauser has more insight as gops contributor. :)

Haven't encountered it so far either, but I think it would be worth to add `SO_REUSEADDR` and `SO_REUSEPORT` to upstream gops to be on the safe side. Will send a PR.

**Comment by borkmann** [comment]
[...]
> Haven't encountered it so far either, but I think it would be worth to add `SO_REUSEADDR` and `SO_REUSEPORT` to upstream gops to be on the safe side. Will send a PR.

Sounds good, thanks @tklauser !

File: operator/main.go [hunk 2]
```diff
@@ -70,10 +70,13 @@ var (
...
-			if err := gops.Listen(gops.Options{}); err != nil {
-				fmt.Fprintf(os.Stderr, "unable to start gops: %s", err)
-				os.Exit(1)
+			// Open socket for using gops to get stacktraces of the agent.
+			addr := fmt.Sprintf("127.0.0.1:%d", viper.GetInt(option.GopsPort))
+			addrField := logrus.Fields{"address": addr}
+			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
+				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
+			} else {
+				log.WithFields(addrField).Info("Started gops server")
```
Comment (tklauser) [inline operator/main.go]: Nit: the `else` branch here is not necessary given that we now `Fatal` on error.

```suggestion
			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
			}
			log.WithFields(addrField).Info("Started gops server")
```

File: daemon/cmd/daemon_main.go [hunk 3]
```diff
@@ -115,9 +115,12 @@ var (
...
 			// Open socket for using gops to get stacktraces of the agent.
-			if err := gops.Listen(gops.Options{}); err != nil {
-				fmt.Fprintf(os.Stderr, "unable to start gops: %s", err)
-				os.Exit(1)
+			addr := fmt.Sprintf("127.0.0.1:%d", viper.GetInt(option.GopsPort))
+			addrField := logrus.Fields{"address": addr}
+			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
+				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
+			} else {
+				log.WithFields(addrField).Info("Started gops server")
```
Comment (tklauser) [inline daemon/cmd/daemon_main.go]: Nit: `else` branch here is not necessary given the `Fatal` above.

File: clustermesh-apiserver/main.go [hunk 4]
```diff
@@ -68,6 +68,15 @@ var (
 		Use:   "clustermesh-apiserver",
 		Short: "Run the ClusterMesh apiserver",
 		Run: func(cmd *cobra.Command, args []string) {
+			// Open socket for using gops to get stacktraces of the agent.
+			addr := fmt.Sprintf("127.0.0.1:%d", viper.GetInt(option.GopsPort))
+			addrField := logrus.Fields{"address": addr}
+			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
+				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
+			} else {
+				log.WithFields(addrField).Info("Started gops server")
```
Comment (tklauser) [inline clustermesh-apiserver/main.go]: Nit: `else` branch here is not necessary given the `Fatal` above.

**Review by tklauser — CHANGES_REQUESTED** [review]
Implementation LGTM apart from the `else` branch nits.

Might be worth adding the new default ports to the table in `Documentation/operations/system_requirements.rst` as well (search for `The following ports should also be available on each node`, cannot link to the source file due to GitHub rendering it).

As already commented, will take care of adding `SO_REUSEADDR`/`SO_REUSEPORT` to `gops` upstream. We can then pull this in a follow-up PR.

File: daemon/cmd/daemon_main.go (hunk 3 above)
Comment (borkmann) [reply daemon/cmd/daemon_main.go]: fixed

File: clustermesh-apiserver/main.go (hunk 4 above)
Comment (borkmann) [reply clustermesh-apiserver/main.go]: fixed

**Comment by borkmann** [comment]
> Implementation LGTM apart from the `else` branch nits.

Done, thanks!

> Might be worth adding the new default ports to the table in `Documentation/operations/system_requirements.rst` as well (search for `The following ports should also be available on each node`, cannot link to the source file due to GitHub rendering it).

Also done now.

> As already commented, will take care of adding `SO_REUSEADDR`/`SO_REUSEPORT` to `gops` upstream. We can then pull this in a follow-up PR.

Added tracking via #14332.

**Comment by borkmann** [comment]
test-me-please

**Review by tklauser — APPROVED** [review]
Thanks!

**Review by glibsm — APPROVED** [review]

**Review by pchaigno — APPROVED** [review]

**Comment by borkmann** [comment]
(4.9 failure potentially related to https://github.com/cilium/cilium/issues/11512)

**Comment by borkmann** [comment]
retest-4.9

**Comment by pchaigno** [comment]
> (4.9 failure potentially related to #11512)

No, I think it was caused by https://github.com/cilium/cilium/pull/14335. Travis CI seems like a flake. Once net-next is green, it should be good to merge :tada:

**Comment by borkmann** [comment]
retest-runtime

**Review by brb — APPROVED** [review]
Good find, LGTM.

**Review by tklauser — COMMENTED** [review]
I think the first two commits should be merged so the version in `go.mod` is bumped as part of the same commit as the `vendor/` directory.

You may also want to add `Fixes #14332` to the commit message bumping `gops`.

Otherwise LGTM, thanks!

**Comment by borkmann** [comment]
> I think the first two commits should be merged so the version in `go.mod` is bumped as part of the same commit as the `vendor/` directory.

I don't always update go.mod, so for me the separation into these two commits was for myself for future reference in that it's clear next time I check the git log that we have go.mod bump as one part and the rest is sync-up.

> You may also want to add `Fixes #14332` to the commit message bumping `gops`.

Done, good point.

**Comment by borkmann** [comment]
test-me-please

**Comment by tklauser** [comment]
> > I think the first two commits should be merged so the version in `go.mod` is bumped as part of the same commit as the `vendor/` directory.
> 
> I don't always update go.mod, so for me the separation into these two commits was for myself for future reference in that it's clear next time I check the git log that we have go.mod bump as one part and the rest is plain sync-up.

The problem with two commits is that it breaks bisects. `go build` will complain about the mismatch between the version in `vendor/` and the one in `go.mod`. What I'd do to bump the dependency is `go get github.com/google/gops@latest && go mod tidy && go mod vendor`. This will update `go.mod`, `go.sum` and `vendor/` and can then be committed as one unit.

But I guess in this case it's also fine to keep the two separate commits for your reference, as bisectability shouldn't really a big issue here. Or maybe adding the command above helps for future reference as well?

**Comment by borkmann** [comment]
> > > I think the first two commits should be merged so the version in `go.mod` is bumped as part of the same commit as the `vendor/` directory.
> > 
> > 
> > I don't always update go.mod, so for me the separation into these two commits was for myself for future reference in that it's clear next time I check the git log that we have go.mod bump as one part and the rest is plain sync-up.
> 
> The problem with two commits is that it breaks bisects. `go build` will complain about the mismatch between the version in `vendor/` and the one in `go.mod`. What I'd do to bump the dependency is `go get github.com/google/gops@latest && go mod tidy && go mod vendor`. This will update `go.mod`, `go.sum` and `vendor/` and can then be committed as one unit.
> 
> But I guess in this case it's also fine to keep the two separate commits for your reference, as bisectability shouldn't really a big issue here. Or maybe adding the command above helps for future reference as well?

Ah, good to know as well. Sure, I can still squash them together after CI is done/green adding the above to the commit decr. That should be fine & we won't need to rerun CI again.

**Review by tklauser — APPROVED** [review]
LGTM with https://github.com/cilium/cilium/pull/14329#issuecomment-742597619, thanks!

**Comment by tklauser** [comment]
Marked for backport to v1.8 as this also fixes an issue with port collision between gops agent and the proxy, see #13400. I'll send a manual backport PR as cherry-picking is not straight forward.

**Comment by tklauser** [comment]
1.8 backport PR: #15634

**Comment by tklauser** [comment]
Apologies, I forgot to mark this as `backport-pending/1.8` when opening #15634. Done so now.

**Pull request merged**